opacus/privacy_engine.py

```diff
@@ -493,3 +495,59 @@ def get_epsilon(self, delta):
             Privacy budget (epsilon) expended so far.
         """
         return self.accountant.get_epsilon(delta)
+
+    def save_checkpoint(
+        self,
+        *,
+        path: Union[str, os.PathLike, BinaryIO, IO[bytes]],
+        module: GradSampleModule,
+        optimizer: Optional[DPOptimizer] = None,
+        noise_scheduler: Optional[_NoiseScheduler] = None,
+        module_state_dict_kwargs: Optional[Dict[str, Any]] = None,
+        torch_save_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Saves the state_dict of module, optimzer, and accountant at path.
+        Args:
+            path: Path to save the state dict objects.
+            module: GradSampleModule to save; wrapped module's state_dict is saved.
+            optimizer: DPOptimizer to save; wrapped optimizer's state_dict is saved.
+            module_state_dict_kwargs: dict of kwargs to pass to ``module.state_dict()``
+            torch_save_kwargs: dict of kwargs to pass to ``torch.save()``
+
+        """
+        dict_to_save = {}
+        dict_to_save["module_state_dict"] = module.state_dict(
+            **(module_state_dict_kwargs or {})
+        )
+        dict_to_save["privacy_accountant_state_dict"] = self.accountant.state_dict()
+        if optimizer is not None:
+            dict_to_save["optimizer_state_dict"] = optimizer.state_dict()
+        if noise_scheduler is not None:
+            dict_to_save["noise_scheduler_state_dict"] = noise_scheduler.state_dict()
+
+        torch.save(dict_to_save, path, **(torch_save_kwargs or {}))
+
+    def load_checkpoint(
```

Comment:
This handles default values beautifully, thanks for that!

Question: do we want to provide a warning in any of the following scenarios:
1) optimizer is provided, but checkpoint doesn't have optimzier_state_dict
2) checkpoint have optimizer_state_dict, but optimizer isn't provided

Reply:
I don't see why not.